### PR TITLE
Run sample program demo scripts in all.sh

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Run sample program demo scripts (`programs/test/*_demo.sh`) in a few components of `all.sh`.

In the long term, we should run the programs in all configurations, but this requires declaring dependencies properly. For the time being, just run the programs in a few builds (out-of-box, full config, with sanitizers, without deprecated functions).

This is a rebase of a part of https://github.com/Mbed-TLS/mbedtls/pull/2698, adapted for the evolution of the development branch and omitting the new demo script for certificate-related programs.

* Create `programs/demo_common.sh` for common logic.
* Create `run_demos.py` and run it from a few components in `all.sh`.
* Do different changes to `check_files.py` because that part of the file has evolved.
* New: Adapt `dlopen_demo.sh` which didn't exist at the time of the original PR. (It's more of a test than a demo, but I prefer to keep things conceptually simple and treat it in the same way.)

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** N/A (only test stuff)
- [x] **backport** https://github.com/Mbed-TLS/mbedtls/pull/7983
- [x] **tests** improving

### CODE REVIEW

File: tests/scripts/run_demos.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python3
+"""Run the Mbed TLS demo scripts.
+"""
+import glob
+import subprocess
+import sys
+
+def run_demo(demo):
+    """Run the specified demo script. Return True if it succeeds."""
+    returncode = subprocess.call([demo])
```
Comment (tom-daubney-arm) [inline tests/scripts/run_demos.py]: I am not sure I understand why `demo` is being passed to `subprocess.call` as an optional parameter?

Comment (gilles-peskine-arm) [reply tests/scripts/run_demos.py]: The first argument to `call` is the program's command line. It's a list where the first element is the program name, which is used as the path to the executable (with `$PATH` lookup) unless there's a separate argument for that.

Comment (tom-daubney-arm) [reply tests/scripts/run_demos.py]: Thanks

**Review by tom-daubney-arm — APPROVED** [review]
I have given this a thorough review to the best of my ability and think it LGTM.
I am not the most experienced with bash however and can't be 100% confident about everything but I have not been able to find any issues with this PR and so I am approving.

It would be really great if the second reviewer was adept at bash to make sure there aren't issues being missed here.

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, ~one non-blocker~ no issues

File: programs/demo_common.sh [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    *) root_dir=".";;
+  esac
+done
+
+## $programs_dir is the directory containing the sample programs.
+# Assume an in-tree build.
+programs_dir="$root_dir/programs"
+
+## msg LINE...
+## msg <TEXT_ORIGIN
```
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply programs/demo_common.sh:46]: Thanks to @tom-cosgrove-arm for pointing out that this is an input redirection, not an example of placeholder syntax.

**Pull request merged**